### PR TITLE
[#2477] Support time-to-live (TTL) header in Kafka-based clients

### LINKED ISSUE #2477: Add time-to-live to Kafka-based Clients
The Kafka-based downstream message APIs already define an optional header *ttl*, but the producer and consumer implementations do not use them by now. The protocol adapters put a TTL property in the metadata (if defined) that is in seconds. The producer should add the creation-time if missing, like it is already done for *ttd* and change the *ttl* from seconds to milliseconds. On the consumer side, records, where the *ttl* elapsed, should be ignored.

### PR DESCRIPTION
**Pull request by b-abel**
This PR adds support for time-to-live (TTL) headers to the Kafka-based clients. It fixes #2477.

### CODE REVIEW

File: clients/adapter-kafka/src/test/java/org/eclipse/hono/adapter/client/telemetry/kafka/AbstractKafkaBasedDownstreamSenderTest.java [hunk 1]
```diff
@@ -335,20 +335,52 @@ public void testThatCreationTimeIsAddedWhenNotPresentAndTtdIsSet(final VertxTest
...
+
+        // WHEN sending the message
+        sender.send(topic, tenant, device, qos, CONTENT_TYPE, null, properties, null)
+                .onComplete(ctx.succeeding(t -> {
+                    ctx.verify(() -> {
+                        // THEN the producer record contains a creation time
+                        final ProducerRecord<String, Buffer> record = mockProducer.history().get(0);
+                        assertThat(record.headers())
+                                .containsOnlyOnce(new RecordHeader("ttl", Json.encode(ttl).getBytes()));
+                        assertThat(record.headers().headers("creation-time")).isNotNull();
```
Comment (sophokles73) [inline clients/adapter-kafka/src/test/java/org/eclipse/hono/adapter/client/telemetry/kafka/AbstractKafkaBasedDownstreamSenderTest.java:362]: is this related to the ttl property?

File: clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/KafkaMessageHelper.java [hunk 2]
```diff
@@ -88,6 +90,37 @@ public static KafkaHeader createKafkaHeader(final String key, final Object value
                 .map(integer -> Integer.valueOf(0).equals(integer) ? QoS.AT_MOST_ONCE : QoS.AT_LEAST_ONCE);
     }
 
+    /**
+     * Gets the {@link MessageHelper#SYS_HEADER_PROPERTY_TTL ttl} header from the given list of Kafka headers.
+     * <p>
+     * If the list contains multiple occurrences of the header, the value of its first occurrence is returned.
+     *
+     * @param headers The headers to get the ttl from.
+     * @return The time-to-live (may be empty).
```
Comment (sophokles73) [inline clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/KafkaMessageHelper.java:99]: this seems to be the wrong text ...

File: clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/consumer/AbstractAtLeastOnceKafkaConsumer.java [hunk 3]
```diff
@@ -230,6 +232,18 @@ private AbstractAtLeastOnceKafkaConsumer(final KafkaConsumer<String, Buffer> kaf
...
 
+    /**
+     * Sets if messages that contain a <em>ttl</em> header where the time-to-live elapsed should be ignored.
+     * <p>
+     * The default is true.
+     *
+     * @param respectTtl The intended behaviour: if true, messages with elapsed ttl are silently dropped and the message
+     *            handler not invoked.
+     */
+    public void setRespectTtl(final boolean respectTtl) {
```
Comment (sophokles73) [inline clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/consumer/AbstractAtLeastOnceKafkaConsumer.java:243]: final ?

File: clients/kafka-common/src/test/java/org/eclipse/hono/client/kafka/consumer/AbstractAtLeastOnceKafkaConsumerKafkaMockConsumerTest.java [hunk 4]
```diff
@@ -266,6 +273,79 @@ public void testThatErrorInMessageHandlerCausesRedelivery(final VertxTestContext
...
+                assertThat(getCommittedOffset(topicPartition)).isEqualTo(1);
+            });
+            ctx.completeNow();
+        };
+
+        // GIVEN a started consumer
+        TestConsumer.createWithMessageHandler(vertxKafkaConsumer, TOPIC, messageHandler).start()
+                .onComplete(ctx.succeeding(v -> {
+                    mockConsumer.updateBeginningOffsets(Map.of(topicPartition, ((long) 0))); // define start offset
+                    mockConsumer.rebalance(Collections.singletonList(topicPartition)); // trigger partition assignment
```
Comment (sophokles73) [inline clients/kafka-common/src/test/java/org/eclipse/hono/client/kafka/consumer/AbstractAtLeastOnceKafkaConsumerKafkaMockConsumerTest.java:301]: maybe on future occasions you might want to use `List.of()` instead ...

File: clients/adapter-kafka/src/test/java/org/eclipse/hono/adapter/client/telemetry/kafka/AbstractKafkaBasedDownstreamSenderTest.java (hunk 1 above)
Comment (b-abel) [reply clients/adapter-kafka/src/test/java/org/eclipse/hono/adapter/client/telemetry/kafka/AbstractKafkaBasedDownstreamSenderTest.java:362]: Yes. Are you asking about the relevance of the `creation-time`? If a `ttl` header is present, we also need the `creation-time` header. `AbstractKafkaBasedDownstreamSender` adds the latter if it's missing. This is what the test verifies.

Comment (sophokles73) [reply clients/adapter-kafka/src/test/java/org/eclipse/hono/adapter/client/telemetry/kafka/AbstractKafkaBasedDownstreamSenderTest.java:362]: Oh yes, I saw that. There also was a comment whether the standard Kafka creation time property should be used (instead of the _creation-time_ property?).

File: clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/KafkaMessageHelper.java [hunk 5]
```diff
@@ -91,12 +91,11 @@ public static KafkaHeader createKafkaHeader(final String key, final Object value
...
-     * Gets the {@link MessageHelper#SYS_HEADER_PROPERTY_TTL ttl} header from the given list of Kafka headers.
-     * <p>
-     * If the list contains multiple occurrences of the header, the value of its first occurrence is returned.
+     * Checks if a {@link MessageHelper#SYS_HEADER_PROPERTY_TTL ttl} header is present and the time already elapsed.
      *
-     * @param headers The headers to get the ttl from.
-     * @return The time-to-live (may be empty).
+     * @param headers The headers to be checked.
+     * @return {@code true} if <em>ttl</em> and <em>creation-time</em> headers are present and the time-to-live is
+     *         already elapsed, {@code false}.
```
Comment (sophokles73) [inline clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/KafkaMessageHelper.java:98]: , {@code false} otherwise ?

File: clients/adapter-kafka/src/test/java/org/eclipse/hono/adapter/client/telemetry/kafka/AbstractKafkaBasedDownstreamSenderTest.java (hunk 1 above)
Comment (b-abel) [reply clients/adapter-kafka/src/test/java/org/eclipse/hono/adapter/client/telemetry/kafka/AbstractKafkaBasedDownstreamSenderTest.java:362]: Oh, yes the TODO :) From what I know, there is quite a bit of complexity in Kafka's handling of message timestamps (behavior can be configured at the topic, timestamps might be changed by the broker in some cases, enabling/disabling of compression affects how timestamps are treated [and so on](https://cwiki.apache.org/confluence/display/KAFKA/KIP-32+-+Add+timestamps+to+Kafka+message). So I tend to use our own header for Hono's messaging and not rely on the message timestamp.

**Review by sophokles73 — APPROVED** [review]
LGTM
